src/devices/multipleAnalogSensors_nws_ros2/Imu_nws_ros2.cpp

```diff
@@ -71,7 +83,20 @@ void Imu_nws_ros2::run()
         imu_ros_data.orientation.x = quat.x();
         imu_ros_data.orientation.y = quat.y();
         imu_ros_data.orientation.z = quat.z();
-        imu_ros_data.orientation.z = quat.w();
+        imu_ros_data.orientation.w = quat.w();
         m_publisher->publish(imu_ros_data);
+
+        if (m_iThreeAxisMagnetometers)
+        {
+           yarp::sig::Vector vecmag(3);
+           m_iThreeAxisMagnetometers->getThreeAxisMagnetometerMeasure(m_sens_index, vecmag, m_timestamp);
```

Comment:
Even if the `m_iThreeAxisMagnetometers` interface exists, it is still possible that `getNrOfThreeAxisMagnetometers()`  returns 0, and this call will return false, so I would either check the return value of `getNrOfThreeAxisMagnetometers()` and or the return value of this function before putting the data in the message.